### ISSUE TITLE
[RFC] Use CSS Custom Properties for dynamic styles

### ISSUE THREAD

**Issue by giuseppeg** [zeit/styled-jsx#595]
The scope of this "RFC" is to discuss about the possibility to use CSS Custom Properties for dynamic styles. 

Switching to use CSS Custom Properties might allow us to explore static extraction in the future.

This mode would be optional and disabled by default as many might still need to support IE, it will also be a breaking change because **we would need to allow interpolations only for CSS values**:

```jsx
// ✅
<style jsx>{`div { color: ${props.color}  }`}</style>
// ✅ the expression below must be hoisted
<style jsx>{`div { color: ${Math.random() > 0.5 ? 'green' : 'black' }  }`}</style>
// 🛑 invalid
<style jsx>{`div { ${props.color ? `color: ${props.color}` : '' }  }`}</style>
``` 

### How dynamic styles work right now</summary>

Interpolations, and therefore the entire template literal containing the CSS, are flagged as dynamic when they have an expression that cannot be evaluated at build time. 

When this is the case we rewrite the styles as follow:

```jsx
<JSXStyle id="123">{`div.__jsx-style-dynamic-selector { color: ${props.color}  }`}</JSXStyle>
```

We also hash the string above to have a base ID, say `123`. Then we rewrite the `className` attribute of every JSX element to include a fn that generates the proper ID at runtime based on props:

```jsx
<div className={JSXStyle.dynamic([  ['123', [props.color] ])}>
   <JSXStyle id="123">{`div.__jsx-style-dynamic-selector { color: ${props.color}  }`}</JSXStyle>
</div>
```

`JSXStyle.dynamic` signature is:

```js
type classNames = string // space separated class names
type JSXStyleInfo = [BaseID, Array<Props>]
type dynamic = (Array<JSXStyleInfo>) => classNames 
```

It takes an `Array<JSXStyleInfo>` because there could be multiple style tags per component.

At runtime then it basically flatten `JSXStyleInfo` and creates an hash from it to finally return classnames

```
jsx-123abc
``` 

Internally then `JSXStyle` does the same and replaces `__jsx-style-dynamic-selector` with `jsx-123abc`.

### Proposal

I think that we could rework this part to instead use custom properties:

We won't use any `__jsx-style-dynamic-selector` but rather use the `BaseID` to generate the scoped class `jsx-123`.

Interpolations would be replaced with custom properties:

```jsx
<JSXStyle id="123">{`div.jsx-123 { color: var(--styled-jsx-123-1)  }`}</JSXStyle>
```

and the root JSXElement would set custom properties using `JSXStyle.dynamic`:

```jsx
<div className="jsx-123" style={{ '--styled-jsx-123-1': props.color }}>
   <JSXStyle id="123">{`div.jsx-123 { color: var(--styled-jsx-123-1)  }`}</JSXStyle>
</div>
```

**Comment by giuseppeg** [comment]
This is just an idea I had a while ago, likely I won't even have time to implement it but I wanted to put it out there in case anybody is interested

cc @developit
